File: packages/node/src/components/pagination.ts
/**
 * Telegram Inline Pagination Keyboard Builder.
 *
 * @packageDocumentation
 */

import type { InlineKeyboardButton, InlineKeyboardMarkup } from "../client/types.js";

/**
 * Customization options for pagination navigation control labels and formatting.
 */
export interface PaginationNavigationLabels {
  /**
   * Label text for the previous page button.
   * Can be a static string or a formatter function `(currentPage, totalPages) => string`.
   * @defaultValue `"Previous"`
   */
  prev?: string | ((currentPage: number, totalPages: number) => string);
  /**
   * Label text for the next page button.
   * Can be a static string or a formatter function `(currentPage, totalPages) => string`.
   * @defaultValue `"Next"`
   */
  next?: string | ((currentPage: number, totalPages: number) => string);
  /**
   * Label text for the middle page indicator button.
   * Can be a static string or a formatter function `(currentPage, totalPages) => string`.
   * @defaultValue `(curr, total) => `${curr} / ${total}``
   */
  pageIndicator?: string | ((currentPage: number, totalPages: number) => string);
  /**
   * Label text displayed when a navigation button is disabled (e.g. on first/last page).
   * @defaultValue `"-"`
   */
  disabled?: string;
  /**
   * Whether to completely hide disabled navigation buttons instead of showing an inactive placeholder.
   * @defaultValue `false`
   */
  hideDisabled?: boolean;
}

/**
 * Options for configuring {@link PaginationKeyboard}.
 *
 * @typeParam T - Type of the item data in the list.
 */
export interface PaginationOptions<T> {
  /** Total items list to paginate over. */
  items: T[];
  /** Current active page index (1-indexed, default: 1). */
  page?: number;
  /** Number of items displayed per page (default: 5). */
  pageSize?: number;
  /** Function returning an {@link InlineKeyboardButton} for each item. */
  itemButton: (item: T, index: number) => InlineKeyboardButton;
  /** Callback data generator for page navigation actions. */
  callbackData?: (action: "prev" | "next" | "noop", page: number) => string;
  /** Custom labels and formatting for navigation buttons. */
  navigation?: PaginationNavigationLabels;
}

/**
 * Fluent builder for creating paginated {@link InlineKeyboardMarkup} menus.
 *
 * @example
 * ```ts
 * const products = ["Item 1", "Item 2", "Item 3", "Item 4", "Item 5", "Item 6"];
 * const keyboard = new PaginationKeyboard({
 *   items: products,
 *   page: 1,
 *   pageSize: 3,
 *   itemButton: (product) => ({ text: product, callback_data: `buy:${product}` }),
 *   callbackData: (action, targetPage) => `page:${targetPage}`,
 *   navigation: {
 *     prev: "Previous",
 *     next: "Next",
 *     pageIndicator: (curr, total) => `Page ${curr} of ${total}`,
 *   },
 * });
 *
 * await bot.sendMessage({
 *   chat_id: 12345,
 *   text: "Store Catalog:",
 *   reply_markup: keyboard.build(),
 * });
 * ```
 */
export class PaginationKeyboard<T = unknown> {
  private readonly items: T[];
  private readonly page: number;
  private readonly pageSize: number;
  private readonly itemButton: (item: T, index: number) => InlineKeyboardButton;
  private readonly callbackData: (action: "prev" | "next" | "noop", page: number) => string;
  private readonly navigation: PaginationNavigationLabels;

  /**
   * Constructs a new {@link PaginationKeyboard} instance.
   *
   * @param options - Configuration options for items, page size, and button mapping.
   */
  constructor(options: PaginationOptions<T>) {
    this.items = options.items;
    this.pageSize = Math.max(1, options.pageSize ?? 5);
    this.page = Math.max(1, options.page ?? 1);
    this.itemButton = options.itemButton;
    this.callbackData =
      options.callbackData ?? ((action, targetPage) => `pagination:${action}:${targetPage}`);
    this.navigation = options.navigation ?? {};
  }

  /**
   * Calculates the total number of pages.
   */
  public get totalPages(): number {
    return Math.max(1, Math.ceil(this.items.length / this.pageSize));
  }

  /**
   * Gets the active page clamped between 1 and totalPages.
   */
  public get currentPage(): number {
    return Math.min(this.page, this.totalPages);
  }

  private resolveLabel(
    label: string | ((curr: number, total: number) => string) | undefined,
    fallback: string | ((curr: number, total: number) => string),
    current: number,
    total: number,
  ): string {
    const target = label ?? fallback;
    return typeof target === "function" ? target(current, total) : target;
  }

  /**
   * Builds the final {@link InlineKeyboardMarkup} structure with item buttons and navigation row.
   *
   * @returns Constructed {@link InlineKeyboardMarkup}.
   */
  public build(): InlineKeyboardMarkup {
    const total = this.totalPages;
    const current = this.currentPage;

    const startIndex = (current - 1) * this.pageSize;
    const pageItems = this.items.slice(startIndex, startIndex + this.pageSize);

    const inline_keyboard: InlineKeyboardButton[][] = pageItems.map((item, idx) => [
      this.itemButton(item, startIndex + idx),
    ]);

    // Navigation control row
    if (total > 1) {
      const navRow: InlineKeyboardButton[] = [];
      const disabledPlaceholder = this.navigation.disabled ?? "-";
      const hideDisabled = this.navigation.hideDisabled ?? false;

      // 1. Previous button
      if (current > 1) {
        const prevText = this.resolveLabel(this.navigation.prev, "Previous", current, total);
        navRow.push({
          text: prevText,
          callback_data: this.callbackData("prev", current - 1),
        });
      } else if (!hideDisabled) {
        navRow.push({
          text: disabledPlaceholder,
          callback_data: this.callbackData("noop", current),
        });
      }

      // 2. Page indicator button
      const indicatorText = this.resolveLabel(
        this.navigation.pageIndicator,
        (curr, tot) => `${curr} / ${tot}`,
        current,
        total,
      );
      navRow.push({
        text: indicatorText,
        callback_data: this.callbackData("noop", current),
      });

      // 3. Next button
      if (current < total) {
        const nextText = this.resolveLabel(this.navigation.next, "Next", current, total);
        navRow.push({
          text: nextText,
          callback_data: this.callbackData("next", current + 1),
        });
      } else if (!hideDisabled) {
        navRow.push({
          text: disabledPlaceholder,
          callback_data: this.callbackData("noop", current),
        });
      }

      inline_keyboard.push(navRow);
    }

    return { inline_keyboard };
  }
}

File: packages/python/src/telebot_py/components/pagination.py
"""Paginated inline keyboard builder (parity with node pagination.ts)."""

from __future__ import annotations

import math
import typing as t
from collections.abc import Callable, Mapping, Sequence

from telebot_py.types import InlineKeyboardButton, InlineKeyboardMarkup

T = t.TypeVar("T")

NavAction = t.Literal["prev", "next", "noop"]
NavLabel = str | Callable[[int, int], str]


class PaginationNavigation(t.TypedDict, total=False):
    """Customization of pagination control labels and formatting.

    Attributes:
        prev: Label of the previous-page button; static or a formatter
            ``(current_page, total_pages) -> label``. Defaults to 'Previous'.
        next: Label of the next-page button, same shapes. Defaults to 'Next'.
        page_indicator: Label of the middle indicator button. Defaults to
            ``"{current} / {total}"``.
        disabled: Placeholder shown for a disabled navigation button.
            Defaults to '-'.
        hide_disabled: Hide disabled navigation buttons entirely instead of
            the placeholder. Defaults to False.
    """

    prev: NavLabel
    next: NavLabel
    page_indicator: NavLabel
    disabled: str
    hide_disabled: bool


def _default_callback_data(action: NavAction, page: int) -> str:
    return f"pagination:{action}:{page}"


def _resolve_label(label: NavLabel | None, fallback: NavLabel, current: int, total: int) -> str:
    target = label if label is not None else fallback
    return target(current, total) if callable(target) else target


class PaginationKeyboard(t.Generic[T]):
    """Fluent builder for paginated inline keyboard menus.

    Example:
        >>> keyboard = PaginationKeyboard(
        ...     items=products,
        ...     page=1,
        ...     page_size=3,
        ...     item_button=lambda item, index: InlineKeyboardButton(
        ...         text=item, callback_data=f"buy:{item}"
        ...     ),
        ... )
        >>> await bot.send_message(chat_id, "Catalog:", reply_markup=keyboard.build())

    Attributes:
        total_pages: Number of pages after chunking ``items`` by page size.
        current_page: The active page, clamped into ``1..total_pages``.
    """

    def __init__(
        self,
        *,
        items: Sequence[T],
        page: int = 1,
        page_size: int = 5,
        item_button: Callable[[T, int], InlineKeyboardButton],
        callback_data: Callable[[NavAction, int], str] | None = None,
        navigation: Mapping[str, t.Any] | None = None,
    ) -> None:
        """Configure the pagination.

        Args:
            items: Total item list to paginate over.
            page: Current active page index (1-indexed).
            page_size: Number of items displayed per page; at least 1.
            item_button: Builds one button per item; receives the item and
                its global index.
            callback_data: Generator for navigation callback data; defaults
                to ``pagination:{action}:{page}``.
            navigation: Custom labels/formatting per
                :class:`PaginationNavigation`.
        """
        self._items = list(items)
        self._page_size = max(1, page_size)
        self._page = max(1, page)
        self._item_button = item_button
        self._callback_data = callback_data if callback_data is not None else _default_callback_data
        self._navigation: Mapping[str, t.Any] = navigation if navigation is not None else {}

    @property
    def total_pages(self) -> int:
        """Number of pages; at least one even for an empty item list."""
        return max(1, math.ceil(len(self._items) / self._page_size))

    @property
    def current_page(self) -> int:
        """The active page clamped between 1 and total_pages."""
        return min(self._page, self.total_pages)

    def build(self) -> InlineKeyboardMarkup:
        """Build the markup with item buttons and the navigation row.

        Returns:
            The constructed InlineKeyboardMarkup; the navigation row is only
            present when there is more than one page.
        """
        total = self.total_pages
        current = self.current_page

        start = (current - 1) * self._page_size
        page_items = self._items[start : start + self._page_size]

        inline_keyboard: list[list[InlineKeyboardButton]] = [
            [self._item_button(item, start + idx)] for idx, item in enumerate(page_items)
        ]

        if total > 1:
            inline_keyboard.append(self._navigation_row(current, total))
        return InlineKeyboardMarkup(inline_keyboard=inline_keyboard)

    def _navigation_row(self, current: int, total: int) -> list[InlineKeyboardButton]:
        """Compose the prev / indicator / next control row.

        Args:
            current: The clamped current page.
            total: Total number of pages.

        Returns:
            The navigation button row.
        """
        nav = self._navigation
        placeholder = t.cast("str", nav.get("disabled", "-"))
        hide_disabled = bool(nav.get("hide_disabled", False))
        row: list[InlineKeyboardButton] = []

        if current > 1:
            row.append(
                InlineKeyboardButton(
                    text=_resolve_label(nav.get("prev"), "Previous", current, total),
                    callback_data=self._callback_data("prev", current - 1),
                )
            )
        elif not hide_disabled:
            row.append(
                InlineKeyboardButton(
                    text=placeholder, callback_data=self._callback_data("noop", current)
                )
            )

        row.append(
            InlineKeyboardButton(
                text=_resolve_label(
                    nav.get("page_indicator"),
                    lambda curr, tot: f"{curr} / {tot}",
                    current,
                    total,
                ),
                callback_data=self._callback_data("noop", current),
            )
        )

        if current < total:
            row.append(
                InlineKeyboardButton(
                    text=_resolve_label(nav.get("next"), "Next", current, total),
                    callback_data=self._callback_data("next", current + 1),
                )
            )
        elif not hide_disabled:
            row.append(
                InlineKeyboardButton(
                    text=placeholder, callback_data=self._callback_data("noop", current)
                )
            )
        return row
